src/main/java/uk/nhs/digital/nhsconnect/nhais/service/RecepConsumerService.java

```diff
@@ -3,67 +3,124 @@
-        long interchangeSequence = recep.getReferenceInterchangeRecep().getInterchangeSequenceNumber();
-
-        for (ReferenceMessageRecep referenceMessageRecep : recep.getReferenceMessageReceps()) {
-            long messageSequence = referenceMessageRecep.getMessageSequenceNumber();
-            ReferenceMessageRecep.RecepCode recepCode = referenceMessageRecep.getRecepCode();
-
-            //TODO could be done in transaction
-            var queryParams = new OutboundStateRepositoryExtensions.UpdateRecepDetailsQueryParams(
-                outbound_sender, outbound_recipient, interchangeSequence, messageSequence);
-            var recepDetails = new OutboundStateRepositoryExtensions.UpdateRecepDetails(
-                recepCode, dateTimePeriod);
-
-            var updatedOutboundState = outboundStateRepository.updateRecepDetails(queryParams, recepDetails);
-            LOGGER.debug("Updated outbound state {} using query {} and recep details {}", updatedOutboundState, queryParams, recepDetails);
-        }
+        var outbound_sender = message.getInterchange().getInterchangeHeader().getRecipient();
+        var outbound_recipient = message.getInterchange().getInterchangeHeader().getSender();
+        var dateTimePeriod = message.getTranslationDateTime().getTimestamp();
+        var interchangeSequence = message.getReferenceInterchangeRecep().getInterchangeSequenceNumber();
+
+        return message.getReferenceMessageReceps().stream()
+            .map(referenceMessageRecep -> prepareOutboundStateUpdateParams(outbound_sender, outbound_recipient, dateTimePeriod, interchangeSequence, referenceMessageRecep))
+            .collect(Collectors.toList());
+    }
+
+    private OutboundStateRepositoryExtensions.UpdateRecepParams prepareOutboundStateUpdateParams(
+        String outbound_sender, String outbound_recipient, Instant dateTimePeriod, Long interchangeSequence, ReferenceMessageRecep referenceMessageRecep) {
+
+        var messageSequence = referenceMessageRecep.getMessageSequenceNumber();
+        var recepCode = referenceMessageRecep.getRecepCode();
+
+        var queryParams = new OutboundStateRepositoryExtensions.UpdateRecepDetailsQueryParams(
+            outbound_sender, outbound_recipient, interchangeSequence, messageSequence);
+        var recepDetails = new OutboundStateRepositoryExtensions.UpdateRecepDetails(
+            recepCode, dateTimePeriod);
+
+        return new OutboundStateRepositoryExtensions.UpdateRecepParams(queryParams, recepDetails);
+    }
+
+    private List<Message> filterOutDuplicates(Interchange interchange) {
+        return interchange.getMessages().stream()
+            .filter(message -> {
+                boolean hasBeenProcessed = hasAlreadyBeenProcessed(message);
+                if (!hasBeenProcessed) {
+                    LOGGER.info("Skipping message {} as it has already been processed", message);
+                }
+                return !hasBeenProcessed;
+            })
```

Comment:
this mean we'll call DB `messssages.size()` times - couldn't we call DB just one time?

Reply:
@mattd-kainos opinion on that? Similar code is in `RegistrationConsumerService`. If we remove that filtering, we could just rely on the unique index and handle issues capturing the unique key violation exception with a warning.

Reply:
@grzegorz-robak @mattd-kainos I suggest we leave it as is and think about this again in https://gpitbjss.atlassian.net/browse/NIAD-358 for potential rework
all ok?

Reply:
I think doing separate inserts is ok. As Bartek said this approach reduces the risk of transmitting duplicates.

Reply:
I'm ok with that. Wanted to be sure we're accepting this.

Reply:
https://gpitbjss.atlassian.net/browse/NIAD-402